// @colyseus/loadtest/lib/rendering.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.error = exports.updateClientsConnected = exports.updateSerializer = exports.render = exports.destroy = exports.setup = void 0;
const util_1 = __importDefault(require("util"));
const blessed_1 = __importDefault(require("blessed"));
const utils_1 = require("./utils");
const packageJson = require(__dirname + "/../package.json");
const RENDER_THROTTLE_MS = 100; // re-render the screen only each 100ms to avoid consuming too much CPU
let screen;
let headerBox;
let clientsBox;
let processingBox;
let networkingBox;
let logBox;
let successfulConnectionBox;
let failedConnectionBox;
let clientsFailed = 0;
let loadTestStartTime = Date.now();
const renderThrottled = utils_1.throttle(() => screen.render(), RENDER_THROTTLE_MS);
function setup(endpoint, roomName, numClients) {
    screen = blessed_1.default.screen({ smartCSR: true });
    headerBox = blessed_1.default.box({
        label: ` ⚔  ${packageJson.name} ${packageJson.version} ⚔  `,
        top: 0,
        left: 0,
        width: "70%",
        height: 'shrink',
        children: [
            blessed_1.default.text({ top: 1, left: 1, tags: true, content: `{yellow-fg}endpoint:{/yellow-fg} ${endpoint}` }),
            blessed_1.default.text({ top: 2, left: 1, tags: true, content: `{yellow-fg}room:{/yellow-fg} ${roomName}` }),
            blessed_1.default.text({ top: 3, left: 1, tags: true, content: `{yellow-fg}serialization method:{/yellow-fg} ...` }),
            blessed_1.default.text({ top: 4, left: 1, tags: true, content: `{yellow-fg}time elapsed:{/yellow-fg} ...` }),
        ],
        border: { type: 'line' },
        style: {
            label: { fg: 'cyan' },
            border: { fg: 'green' }
        }
    });
    successfulConnectionBox = blessed_1.default.text({ top: 2, left: 1, tags: true, content: `{yellow-fg}connected:{/yellow-fg} 0` });
    failedConnectionBox = blessed_1.default.text({ top: 3, left: 1, tags: true, content: `{yellow-fg}failed:{/yellow-fg} 0` });
    clientsBox = blessed_1.default.box({
        label: ' clients ',
        left: "70%",
        width: "30%",
        height: 'shrink',
        children: [
            blessed_1.default.text({ top: 1, left: 1, tags: true, content: `{yellow-fg}numClients:{/yellow-fg} ${numClients}` }),
            successfulConnectionBox,
            failedConnectionBox
        ],
        border: { type: 'line' },
        tags: true,
        style: {
            label: { fg: 'cyan' },
            border: { fg: 'green' },
        }
    });
    processingBox = blessed_1.default.box({
        label: ' processing ',
        top: 6,
        left: "70%",
        width: "30%",
        height: 'shrink',
        border: { type: 'line' },
        children: [
            blessed_1.default.text({ top: 1, left: 1, tags: true, content: `{yellow-fg}memory:{/yellow-fg} ...` }),
            blessed_1.default.text({ top: 2, left: 1, tags: true, content: `{yellow-fg}cpu:{/yellow-fg} ...` }),
        ],
        tags: true,
        style: {
            label: { fg: 'cyan' },
            border: { fg: 'green' },
        }
    });
    networkingBox = blessed_1.default.box({
        label: ' networking ',
        top: 11,
        left: "70%",
        width: "30%",
        border: { type: 'line' },
        children: [
            blessed_1.default.text({ top: 1, left: 1, tags: true, content: `{yellow-fg}bytes received:{/yellow-fg} ...` }),
            blessed_1.default.text({ top: 2, left: 1, tags: true, content: `{yellow-fg}bytes sent:{/yellow-fg} ...` }),
        ],
        tags: true,
        style: {
            label: { fg: 'cyan' },
            border: { fg: 'green' },
        }
    });
    logBox = blessed_1.default.box({
        label: ' logs ',
        top: 7,
        width: "70%",
        padding: 1,
        border: { type: 'line' },
        tags: true,
        style: {
            label: { fg: 'cyan' },
            border: { fg: 'green' },
        },
        // scroll
        scrollable: true,
        input: true,
        alwaysScroll: true,
        scrollbar: {
            style: {
                bg: "green"
            },
            track: {
                bg: "gray"
            }
        },
        keys: true,
        vi: true,
        mouse: true
    });
    // Trap built-in console log methods.
    console.log = function (...args) {
        logBox.content = args.map(arg => util_1.default.inspect(arg)).join(" ") + "\n" + logBox.content;
        renderThrottled();
    };
    console.warn = function (...args) {
        logBox.content = `{yellow-fg}${args.map(arg => util_1.default.inspect(arg)).join(" ")}{/yellow-fg}\n${logBox.content}`;
        renderThrottled();
    };
    console.error = function (...args) {
        logBox.content = `{red-fg}${args.map(arg => util_1.default.inspect(arg)).join(" ")}{/red-fg}\n${logBox.content}`;
        renderThrottled();
    };
    // append widgets to the screen.
    screen.key(['escape', 'q', 'C-c'], (ch, key) => process.exit(0)); // Quit on Escape, q, or Control-C.
    screen.title = "@colyseus/loadtest";
    screen.append(headerBox);
    screen.append(clientsBox);
    screen.append(logBox);
    screen.append(processingBox);
    screen.append(networkingBox);
    screen.render();
}
exports.setup = setup;
function destroy() {
    screen.destroy();
}
exports.destroy = destroy;
/**
 * Update memory / cpu usage
 */
let startTime = process.hrtime();
let startUsage = process.cpuUsage();
function render(stats) {
    /**
     * Program elapsed time
     */
    const elapsedTimeText = headerBox.children[3];
    elapsedTimeText.content = `{yellow-fg}time elapsed:{/yellow-fg} ${utils_1.elapsedTime(Math.round((Date.now() - loadTestStartTime) / 1000))}`;
    /**
     * Memory / CPU Usage
     */
    const memoryText = processingBox.children[0];
    memoryText.content = `{yellow-fg}memory:{/yellow-fg} ${(process.memoryUsage().heapUsed / 1024 / 1024).toFixed(2)} MB`;
    var elapTime = process.hrtime(startTime);
    var elapUsage = process.cpuUsage(startUsage);
    var elapTimeMS = elapTime[0] * 1000 + elapTime[1] / 1000000;
    var elapUserMS = elapUsage.user / 1000;
    var elapSystMS = elapUsage.system / 1000;
    var cpuPercent = (100 * (elapUserMS + elapSystMS) / elapTimeMS).toFixed(1);
    const cpuText = processingBox.children[1];
    cpuText.content = `{yellow-fg}cpu:{/yellow-fg} ${cpuPercent}%`;
    renderThrottled();
    startTime = process.hrtime();
    startUsage = process.cpuUsage();
    /**
     * Networking
     */
    const bytesReceivedBox = networkingBox.children[0];
    bytesReceivedBox.content = `{yellow-fg}bytes received:{/yellow-fg} ${utils_1.formatBytes(stats.bytesReceived)}`;
    const bytesSentBox = networkingBox.children[1];
    bytesSentBox.content = `{yellow-fg}bytes sent:{/yellow-fg} ${utils_1.formatBytes(stats.bytesSent)}`;
    updateClientsConnected(stats.clientsConnected);
}
exports.render = render;
function updateSerializer(serializerId) {
    // display serialization method in the UI
    const serializerIdText = headerBox.children[2];
    serializerIdText.content = `{yellow-fg}serialization method:{/yellow-fg} ${serializerId}`;
}
exports.updateSerializer = updateSerializer;
function updateClientsConnected(clientsConnected) {
    successfulConnectionBox.content = `{yellow-fg}connected:{/yellow-fg} ${clientsConnected}`;
    renderThrottled();
}
exports.updateClientsConnected = updateClientsConnected;
function error(message) {
    console.error(message);
    clientsFailed++;
    failedConnectionBox.content = `{red-fg}failed:{/red-fg} ${clientsFailed}`;
    renderThrottled();
}
exports.error = error;
